### ISSUE TITLE
where(obj) with json fields

### ISSUE THREAD

**Issue by kapouer** [Vincit/objection.js#348]
Do objection supports
```
Model.query().where({
  "tab.col:field.subfield": "val"
})
```
?

**Comment by kapouer** [comment]
I found two other ways which involve not using objects:
```
whereJsonField("tab.col:field.subfield", "=", "val")
```
which is a bit verbose because of the need to use the "=" operator,
and also `whereJsonField("tab.othercol", "otherval")` produces invalid sql by trying to access a json without path.

and
```
where(ref("tab.col:field.subfield").castText(), "val")
```
which could be a bit smarter about the type to default type to cast to (text because "val" is a string),
but doesn't work within an object:

```
where({
  [ref("tab.col:field.subfield").castText()]: "val"
})
```
the reference is just passed as is to knex, apparently.

**Comment by koskimas** [comment]
Default casting could be improved i quess. @elhigu could what do you think? Is it feasible?

```js
where({
  [ref("tab.col:field.subfield").castText()]: "val"
})
```

This just creates an object `{"[Object object]": "val"}` since `ref().castText()` returns an instance of `ReferenceBuilder`.

**Comment by fl0w** [comment]
Wouldn't a `ReferenceBuilder#toString()` implementation solve this?

**Comment by koskimas** [comment]
No it wouldn't. The whole point of `ref` function is that you don't need to access it through the knex instance (ref doesn't care which database is used). The `ref` instance is then expanded to SQL in build phase where a knex instance is available. We could simply return the input string from `toString()` but then why use `ref`at all?

**Comment by fl0w** [comment]
Ah, I understand. Apologies.

**Comment by elhigu** [comment]
@kapouer currently the correct way to do this is:

```js
.where(ref("tab.col:field.subfield").castText(), "val")
```

Old way with`.whereJsonField(...)` should be deprecated. I haven't added (at least intentionally) support to `.where(ojb)` to refer json attributes directly (I'm not sure if even field expressions can be used there yet to refer json attributes). 

If the support is implemented one might need to do this (because string format of field expression doesn't support any casting operations):

```js
.where({
  'col:myJsonAttr' : raw('?::jsonb', [val])
})
```

or if `lit()` is implemented:

```js
.where({
  'col:myJsonAttr' : lit("val").castJson()
})
```

**Comment by koskimas** [comment]
Implementing this

```js
.where({
  'col:myJsonAttr' : ...
})
```

would completely rule out using `:` in a column name with objection since `:` is the only thing that can be used to detect a field expression. I don't think we should do that.

Automatic casting could be done for strings and booleans but that would once again require that objection parses the where clauses and duplicates the logic from knex. Automatic casting for various numeric types would be more difficult or impossible.

`lit` is not implemented and after 0.9 release you can do this:

```js
.where(ref("tab.col:field.subfield"), lit("val").castJson())
```

or the this.

```js
.where(ref("tab.col:field.subfield").castText(), "val")
```

I think that should be good enough for now.
